examples/LabeledModels/LabeledModels.py

```diff
@@ -0,0 +1,564 @@
+#
+
+# %%
+dataset = Dataset(
+    {
+        "cNrm": cNrm,
+        "v": v,
+        "v_der": v_der,
+        "v_inv": v_inv,
+        "v_der_inv": v_der_inv,
+    }
+)
+dataset
+
+# %% [markdown]
+# We can also create separate datasets for the value function variables and the policy function variable.
+#
+
+# %%
+value_function = Dataset(
+    {"v": v, "v_der": v_der, "v_inv": v_inv, "v_der_inv": v_der_inv}
+)
+policy_function = Dataset({"cNrm": cNrm})
+
+# %% [markdown]
+# Up to now I've used the word function for the variables stored as datasets. This is because using the `interp` method we can interpolate the values of the variables at any point in the state space. This is useful for solving the model, as we will see later.
+#
+
+# %%
+dataset.interp({"mNrm": np.sort(np.random.uniform(epsilon, 20, 10))})
+
+
+# %% [markdown]
+# Because of the curvature of the value and marginal value functions, it'll be useful to use the inverse value and marginal value functions instead and re-curve them. For this, I create a new class `ValueFunctionCRRA` that returns the appropriate value and marginal value functions.
+#
+
+# %%
+class ValueFunctionCRRA(object):
+    def __init__(self, dataset: xr.Dataset, CRRA: float):
+
+        self.dataset = dataset
+        self.CRRA = CRRA
+
+    def __call__(self, state):
+        """
+        Interpolate inverse falue function then invert to get value function at given state.
+        """
+
+        result = CRRAutility(
+            self.dataset["v_inv"].interp(
+                state,
+                assume_sorted=True,
+                kwargs={"fill_value": "extrapolate"},
+            ),
+            self.CRRA,
+        )
+
+        result.name = "v"
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+    def derivative(self, state):
+        """
+        Interpolate inverse marginal value function then invert to get marginal value function at given state.
+        """
+        result = CRRAutilityP(
+            self.dataset["v_der_inv"].interp(
+                state,
+                assume_sorted=True,
+                kwargs={"fill_value": "extrapolate"},
+            ),
+            self.CRRA,
+        )
+
+        result.name = "v_der"
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+    def evaluate(self, state):
+        """
+        Interpolate all data variables in the dataset.
+        """
+
+        result = self.dataset.interp(state, kwargs={"fill_value": "extrapolate"})
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+
+# %%
+vfunc = ValueFunctionCRRA(value_function, CRRA)
+
+# %%
+rand_states = np.sort(np.random.uniform(mVec[1], mVec[-1], 100))
+rand_states
+
+# %%
+rand_ds = vfunc.evaluate({"mNrm": rand_states})
+rand_ds
+
+# %%
+rand_v = vfunc({"mNrm": rand_states})
+np.max(np.abs(rand_v - rand_ds["v"]))
+
+
+# %% [markdown]
+# ### State transitions
+#
+
+# %%
+def state_transition(s=None, a=None, params=None):
```

Comment:
Ah, yes. I've found it useful to write transition functions in similar way.

In my code, I've been using `x` for states and `k` for discretely realized values of shocks.

Reply:
Here, s=state, a=action(s). Elsewhere I might have used `shocks` explicitly, so I should probably be more consistent with this nomenclature.